idaes/core/scaling/autoscaling.py

```diff
@@ -0,0 +1,291 @@
+              (default=True)
+
+        Returns:
+            None
+
+        Raises:
+            TypeError is blk_or_cons is not a Block or Constraint
+            ValueError if norm is not a positive integer
+        """
+        # Validate norm
+        # First cast to int
+        norm = int(norm)
+        if norm < 1:
+            raise ValueError(
+                f"Invalid value for norm in constraints_by_jacobian_norm ({norm}). "
+                "Value must be a positive integer."
+            )
+
+        if isinstance(blk_or_cons, BlockData):
+            # Scalar Block or element of Indexed Block
+            if descend_into:
+                # Scale all constraints, so pass con_list=None
+                con_list = None
+            else:
+                # Otherwise, get an iterator of constraints only on block
+                con_list = blk_or_cons.component_data_objects(
+                    Constraint, descend_into=False
+                )
+            self._con_by_norm(blk_or_cons, con_list=con_list, norm=norm)
+
+        elif isinstance(blk_or_cons, Block):
+            # Indexed Block
+            for b in blk_or_cons.values():
+                if descend_into:
+                    # Scale all constraints, so pass con_list=None
+                    con_list = None
+                else:
+                    # Otherwise, get an iterator of constraints only on block
+                    con_list = b.component_data_objects(Constraint, descend_into=False)
+                self._con_by_norm(b, con_list=con_list, norm=norm)
+
+        elif isinstance(blk_or_cons, ConstraintData):
+            # Scalar Constraint or element of Indexed Constraint
+            self._con_by_norm(
+                blk_or_cons.parent_block(), con_list=[blk_or_cons], norm=norm
+            )
+
+        elif isinstance(blk_or_cons, Constraint):
+            # Indexed Constraint
+            self._con_by_norm(
+                blk_or_cons.parent_block(), con_list=blk_or_cons.values(), norm=norm
+            )
+
+        else:
+            raise TypeError(f"{blk_or_cons.name} is not a block or constraint.")
+
+    def _vardata_by_magnitude(self, vardata):
+        if vardata.fixed:
+            # Fixed var
+            return
+
+        if value(vardata) is None:
+            sf = 1
+        else:
+            val = abs(value(vardata))
+            if val <= self.config.zero_tolerance:
+                sf = 1
+            else:
+                sf = 1 / val
+
+        self._set_scaling_factor(vardata, "variable", sf)
+
+    def _con_by_norm(self, blk, con_list=None, norm=2):
+        # Get scaled Jacobian - we want to consider any existing scaling
+        # We will account for existing scaling factors later and update them
+        jac, nlp = get_jacobian(blk, scaled=True)
+
+        if con_list is None:
+            con_list = nlp.get_pyomo_equality_constraints()
```

Comment:
In my experience, `get_jacobian` is slow because forming the Pynumero NLP is slow. If we have flowsheet-level variables, we end up getting a Jacobian for the flowsheet and every subcomponent. I'm not sure how slow forming the NLP is going to be for PropertyBlock 43 in the MEA column, but you're still at least doubling the work you're doing.

I would suggest getting the Jacobian and NLP once at the beginning of `constraints_by_jacobian_norm`, then just referring to rows and columns of the overall object as you want to scale constraints.

Reply:
I have been thinking about there best way to do this, and did realize I was probably getting the Jacobian more times than necessary. I was not aware of the issue of having higher-level components in the problem causing the entire problem to be emitted (and given how properties work that will be an issue). I am also thinking there might need to be some checks to deal with deactivated components (and that it might be possible to deactivate unnecessary components before getting the Jacobian to reduce the problem size).

Reply:
I think I've fixed this so this method will only be called once for any call to the scaling method.

`constraints_by_jacobian_norm` (line 156) collects all the constraints we want to scale and the top level block, and then only calls the autoscaler once.